src/server/game/Entities/Player/Player.cpp

```diff
@@ -22138,6 +22138,35 @@ void Player::SendCooldownEvent(SpellInfo const* spellInfo, uint32 itemId /*= 0*/
     data << uint32(spellInfo->Id);
     data << uint64(GetGUID());
     SendDirectMessage(&data);
+    
+    uint32 cat = spellInfo->GetCategory();
+    uint32 catRec = spellInfo->CategoryRecoveryTime;
+    if (cat && catRec)
+    {
+        SpellCategoryStore::const_iterator ct = sSpellsByCategoryStore.find(cat);
+        if (ct != sSpellsByCategoryStore.end())
```

Comment:
Dont know how it should be done, but it looks pretty ugly.

If i got it right, the code searches all spells in dbc with same category for every spell used by a player.
Then inside every category it searches if any of those spells has a "cooldown" stored in  the player.

Imagine how many times this double loop can be executed...